crates/ide-assists/src/handlers/wrap_return_type_in_result.rs

```diff
@@ -1,16 +1,77 @@
 use std::iter;
 
+use hir::HasSource;
 use ide_db::{
     famous_defs::FamousDefs,
     syntax_helpers::node_ext::{for_each_tail_expr, walk_expr},
 };
+use itertools::Itertools;
 use syntax::{
-    ast::{self, make, Expr},
-    match_ast, ted, AstNode,
+    ast::{self, make, Expr, HasGenericParams},
+    match_ast, ted, AstNode, ToSmolStr,
 };
 
 use crate::{AssistContext, AssistId, AssistKind, Assists};
 
+fn result_ty(ctx: &AssistContext<'_>, ret_type: &ast::Type) -> Option<ast::Type> {
+    let scope = ctx.sema.scope(ret_type.syntax())?;
+    let core_result = FamousDefs(&ctx.sema, scope.krate()).core_result_Result()?;
+
+    let ty = ctx.sema.resolve_type(ret_type)?.as_adt();
+    if matches!(ty, Some(hir::Adt::Enum(ret_type)) if ret_type == core_result) {
+        // The return type is already wrapped in a Result
+        cov_mark::hit!(wrap_return_type_in_result_simple_return_type_already_result);
+        return None;
+    }
+
+    // Try to find a Result<T, ...> type alias in the current scope (shadowing the default).
+    let result_name = core_result.name(ctx.db());
+    let mut local_result = None;
+    scope.process_all_names(&mut |name, def| match def {
+        hir::ScopeDef::ModuleDef(hir::ModuleDef::TypeAlias(alias)) if name == result_name => {
+            let Some(enum_ty) = alias.ty(ctx.db()).as_adt().and_then(|adt| adt.as_enum()) else {
+                return;
+            };
+
+            let ret_is_alias = matches!(ty, Some(hir::Adt::Enum(ret_type)) if ret_type == enum_ty);
+            if enum_ty != core_result || ret_is_alias {
+                // The current name is not a Result alias, or the return type is already this Result alias.
```

Comment:
We cannot stop looping in the second case since `SemanticScope::process_all_names` doesn't accept `ControlFlow` or return an iterator. This is a bit wasteful but I'm sure if its worth modifying the API, thoughts?

There's a good chance this is already possible with different API I missed, but I figured I might as well ask :sweat_smile:

Reply:
I think it'd be nicer for us to expose a `resolve_mod_path` function on `Semantics` that takes a `ModPath` to resolve, then we can just call that with a `ModPath::from_segmenst(Plain, iter;:once(sym::Result))`

Reply:
Sounds good, I've pushed a fix to do so. I'm not entirely sure if the signature is sensible as of now though.